### DISCUSSION TITLE
conflicts found when running AI-ODMR and Instream

### DISCUSSION THREAD

**Discussion by liuyanulm** [Ulm-IQO/qudi-core#73] [Q&A] [answered]
I am running NV ensemble ODMR with a photodiode where one of the AI channel is used to measure the voltage ouput of the photodiode. While an error report occured with the AI Instream. The used NI card is USB-6343.

Endeavours so far:
1. I am not familiar with the Instream. But if I remember correctly, ODMR logic is just stealing Counter 2 as clock channel and Counter3 as photon counter from the confocal scanlogic. An additional AI counting (scann_ai_counter) should not take more NI resources from the Instream. However, the error code according to NI website, is most likely associated with counter resources. In case there is not enough counters for intreaming, since with this setup confocal scan is not used I even freed all AOs, Counter0 and Counter1 from the NI hardware file, the same error continued to happen.
2. changing AIs channels have been tested. Only resulted in some different error reports.

How to reproduce this bug:
Only need to add ai in your NI/ODMR module.

**Comment by liuyanulm** [comment]
If the streamer and the odmr is using the same ai source, I got the error code:
`PyDAQmx.DAQmxFunctions.PALResourceReservedError: The specified resource is reserved. The operation could not be completed as specified.
Task Name: ODMRAnalog

Status Code: -50103
 in function DAQmxStartTask`

**Reply by liuyanulm** [reply]
By this error report, perhaps the conflict has nothing to do with the counter resources, but with how the analog in task is handled in the NI hardware files.

**Reply by liuyanulm** [reply]
Examine the counter resources again,
`nicard_6343_instreamer._clk_task_handle.channel_names[0]
Out[5]: 'Dev2/ctr0'

nicard_6343._scanner_clock_channel
Out[6]: '/Dev2/Ctr2'`

Now it is quite clear that the conflict is with the Analog-in handling.

**Comment by liuyanulm** [comment]
> If the streamer and the odmr is using the same ai source, I got the error code: `PyDAQmx.DAQmxFunctions.PALResourceReservedError: The specified resource is reserved. The operation could not be completed as specified. Task Name: ODMRAnalog
> 
> Status Code: -50103 in function DAQmxStartTask`



> If the streamer and the odmr is using the same ai source, I got the error code: `PyDAQmx.DAQmxFunctions.PALResourceReservedError: The specified resource is reserved. The operation could not be completed as specified. Task Name: ODMRAnalog
> 
> Status Code: -50103 in function DAQmxStartTask`

Additional instruction, this is the error report, when running ODMR after start the Instreamer.
If Instreamer is started after ODMR running, the error reports:
'nidaqmx.errors.DaqError: The specified resource is reserved. The operation could not be completed as specified.
Task Name: AnalogIn_2823682196104

Status Code: -50103
Error while starting streaming device data acquisition.
'

**Comment by liuyanulm** [comment]
A new test with Instreamer and confocal scan.
With AI-Instreamer running, to start a confocal scan, error report:
"PyDAQmx.DAQmxFunctions.PALResourceReservedError: The specified resource is reserved. The operation could not be completed as specified.
Task Name: ScanAnalogIn" 

Since the scanner logic and odmrlogic are sharing the same scanning functions in the nicard6343 hardware file, this also suggests that the AnalogIn daqtask in the hardware file possibly have some problems.

**Comment by liuyanulm** [comment, marked as answer]
I contacted an engineer from NI. 
I was told that, all AnalogIn channels in one single NI card only supports one sample rate.
I guess this is the answer...


However, it is possible to do it in software. For example, when odmr or confocal is running with one ai channel, the counter logic or instream logic can calculate an coefficient between the streaming clock and scanner clock, and then stream the ai values with the accordingly. Even the streamed ai values are not directly from the ni card, but it would still be helpful to users. Additionally, we can also use a checkbox for users to choose the prefered data source, accutual measured intensities or revised data from scanners.

**Reply by Neverhorst** [reply]
Yes, that is how the analog-in in these cards work (it's a single ADC for all channels combined... cheap). There has been the concept around for quite some time, to write a class that controls and governs a single NI card and provides functionality to several qudi hardware modules (e.g. InStreamer, Confocal etc.) via import. However so far the pain has not been great enough for someone to do that. It's also a rather non-trivial task to do that properly.

**Comment by liuyanulm** [comment]
> Yes, that is how the analog-in in these cards work (it's a single ADC for all channels combined... cheap). There has been the concept around for quite some time, to write a class that controls and governs a single NI card and provides functionality to several qudi hardware modules (e.g. InStreamer, Confocal etc.) via import. However so far the pain has not been great enough for someone to do that. It's also a rather non-trivial task to do that properly.

I got an idea of how to produce AI counts for both InStream and the Scanner at the same time. Run AI counting always with the highest sample rate (500kSample/s), and then averaging the AI values by the counting period of InStream and the Scanner respectively (trigger start of the data handling by software).

**Reply by liuyanulm** [reply]
This probably requires an additional AI hardware class, and revision of the AI functions in other NI card hardware classes to not to start AI daqtask directly instead taking the AI meta data from this AI hardware class.

**Reply by Neverhorst** [reply]
What you describe is exactly the "governing class" I was talking about in my earlier comment. 😊 
Does not neccessarily need to be a qudi hardware module. A normal Python class that can be imported and handles NI devices globally should also be enough. The main challenge is exception and conflict handling in the actual hardware modules that use this shared resource.

**Reply by liuyanulm** [reply]
Not exactly. Your "governing class" looks a lot bigger and the data handling seems quite challenging. I was proposing only to separate AI tasks from the current Instream and scanners to avoid conflict. The new AI class do AI tasks direct with NI card and should only reads AI values of max sampling rate from configuered channels, and pass AI values to Instream and scanner, the other two does the AI averaging according to their own streaming clock and scanner clock.
the class can be named with "AI_reader" by the way😊
